crates/matrix-sdk-ui/tests/integration/sync_service.rs

```diff
@@ -202,3 +205,123 @@ async fn test_sync_service_state() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+#[async_test]
+async fn test_sync_service_offline_mode() {
+    let mock_server = MatrixMockServer::new().await;
+    let client = mock_server.client_builder().build().await;
+
+    let sync_service = SyncService::builder(client).with_offline_mode().build().await.unwrap();
+    let mut states = sync_service.state();
+
+    Mock::given(SlidingSyncMatcher)
+        .respond_with(ResponseTemplate::new(404))
+        .expect(1..)
+        .mount(mock_server.server())
+        .await;
+
+    {
+        let _versions_guard = mock_server.mock_versions().error500().mount_as_scoped().await;
+
+        sync_service.start().await.expect("We should be able to start the sync service");
+        assert_next_eq!(states, State::Running);
+
+        let next_state = timeout(states.next(), Duration::from_millis(500))
+            .await
+            .expect("We should have changed states by now")
+            .unwrap();
+
+        assert_eq!(next_state, State::Offline, "We should have entered the offline mode");
```

Comment:
we do have the rather verbosely named `assert_next_with_timeout` helper that combines both checks, fwiw.

Reply:
Ah I was searching for something like this, though that's still not quite ideal so I added a new helper: 0010d37a55f686014926f997f48bf6167e530f20.